### ISSUE TITLE
Failed to register the resource using the endpoint `POST /resource/{repository}/{type}/{tag}`

### ISSUE THREAD

**Issue by yaoxin1995** [confidential-containers/kbs#69]
Hey, 

I tried to register resources to KBS using the HTTP endpoint that KBS has defined. 

But I got the following error:
```
HTTP/1.1 405 Method Not Allowed
content-length: 0
allow: GET
date: Fri, 31 Mar 2023 08:34:18 GMT
Proxy-Connection: keep-alive
Connection: keep-alive
```

How I send the request:
```
curl -X POST http://$1/kbs/v0/resource/default/test_resources/test \
     -i \
     -H 'Content-Type: text/plain' \
     -d ewogICAgInRlZS1wdWJrZXkiOiB7Cgkia3R5IjogImtleV90eXBlIiwKCSJhbGciOiAiYWxnb3JpdGhtIiwKCSJrIjogIm15X2tleSIKICAgIH0sCiAgICAidGVlLWV2aWRlbmNlIjogIm15X2V2aWRlbmNlIgp9Cg==



```
On server side, I got : 
```
[2023-03-31T11:23:45Z INFO  actix_web::middleware::logger] 127.0.0.1 "POST /kbs/v0/resource/default/test_resources/test HTTP/1.1" 405 0 "-" "curl/7.68.0" 0.000371
```

The server did have defined the post method for this endpoint, as you can see:
```
        let http_server = HttpServer::new(move || {
            App::new()
                .wrap(middleware::Logger::default())
                .app_data(web::Data::clone(&sessions))
                .app_data(web::Data::clone(&attestation_service))
                .app_data(web::Data::new(repository.clone()))
                .app_data(web::Data::new(http_timeout))
                .app_data(web::Data::new(user_public_key.clone()))
                .app_data(web::Data::new(insecure_api))
                .service(web::resource(kbs_path!("auth")).route(web::post().to(http::auth)))
                .service(web::resource(kbs_path!("attest")).route(web::post().to(http::attest)))
                .service(
                    web::resource([
                        kbs_path!("resource/{repository}/{type}/{tag}"),
                        kbs_path!("resource/{type}/{tag}"),
                    ])
                    .route(web::get().to(http::get_resource)),
                )
                .service(
                    web::resource([
                        kbs_path!("resource/{repository}/{type}/{tag}"),
                        kbs_path!("resource/{type}/{tag}"),
                    ])
                    .route(web::post().to(http::set_resource)),
                )
                .service(web::resource([
                    kbs_path!("hello"),
                ])
                .route(web::get().to(http::get_hello)))
        });
```

Do you know how I can register the resource correctly?

**Comment by jialez0** [comment]
@yaoxin1995 Registering resources requires authentication, so there is the way to solve this problem:

Generate a user auth key pair:
```shell
openssl genpkey -algorithm ed25519 > private.key
openssl pkey -in config/private.key -pubout -out public.pub
```
Start KBS with the user's public key:
```shell
kbs --socket 127.0.0.1:8080 --auth-public-key $(realpath public.pub)
```
Then you should generate a JWT signed with the private key, and take the token with POST request, then you can register the resource correctly.

Or, you can start KBS with `--insecure-api` arg like:
```shell
kbs --socket 127.0.0.1:8080 --insecure-api
```
Then you can register the resource without a JWT.

**Comment by yaoxin1995** [comment]
> curl -X POST http://$1/kbs/v0/resource/default/test_resources/test \
>      -i \
>      -H 'Content-Type: text/plain' \
>      -d ewogICAgInRlZS1wdWJrZXkiOiB7Cgkia3R5IjogImtleV90eXBlIiwKCSJhbGciOiAiYWxnb3JpdGhtIiwKCSJrIjogIm15X2tleSIKICAgIH0sCiAgICAidGVlLWV2aWRlbmNlIjogIm15X2V2aWRlbmNlIgp9Cg==

@jialez0  Thanks. 

But still, I can't register resources to kbs.  

I have disabled the `https` and `jwt` as follows:
```
sudo ./target/debug/kbs --socket 0.0.0.0:8000 --insecure-http --insecure-api

```
Then I use the `curl post` to register the resource:
```
curl -X POST http://localhost:8000/kbs/v0/resource/default/test_resources/test \
     -i \
     -H 'Content-Type: text/plain' \
     -d ewogICAgInRlZS1wdWJrZXkiOiB7Cgkia3R5IjogImtleV90eXBlIiwKCSJhbGciOiAiYWxnb3JpdGhtIiwKCSJrIjogIm15X2tleSIKICAgIH0sCiAgICAidGVlLWV2aWRlbmNlIjogIm15X2V2aWRlbmNlIgp9Cg==

```

I got the `method not allow response` :
```
HTTP/1.1 405 Method Not Allowed
content-length: 0
allow: GET
date: Sun, 02 Apr 2023 08:44:00 GMT
Proxy-Connection: keep-alive
Connection: keep-alive

```
As you can see, the `/kbs/v0/resource/{repo}/{type}/{tag}` only allows the `get` method.

On the server side, it can receive the `curl post` request:
```
[2023-04-02T08:44:00Z INFO  actix_web::middleware::logger] 127.0.0.1 "POST /kbs/v0/resource/default/test_resources/test HTTP/1.1" 405 0 "-" "curl/7.68.0" 0.000384
```

**Comment by jialez0** [comment]
@Xynnn007 Could you please take a look at this issue? :-)

**Comment by Xynnn007** [comment]
Hi @yaoxin1995 it works on my side.

From my side, KBS
```
root@xynnn-hongkong-1:/home/kbs# ./target/debug/kbs --socket 0.0.0.0:8000 --insecure-http --insecure-api
[2023-04-02T10:30:03Z WARN  kbs] insecure apis are enabled.
[2023-04-02T10:30:03Z INFO  api_server] Starting HTTP server at [0.0.0.0:8000]
[2023-04-02T10:30:03Z INFO  actix_server::builder] starting 4 workers
[2023-04-02T10:30:03Z INFO  actix_server::server] Tokio runtime found; starting in existing Tokio runtime
[2023-04-02T10:30:07Z INFO  actix_web::middleware::logger] 127.0.0.1 "POST /kbs/v0/resource/default/test_resources/test HTTP/1.1" 200 0 "-" "curl/7.68.0" 0.000941
```

and `curl`
```
curl -X POST http://localhost:8000/kbs/v0/resource/default/test_resources/test      -i      -H 'Content-Type: text/plain'      -d ewogICAgInRlZS1wdWJrZXkiOiB7Cgkia3R5IjogImtleV90eXBlIiwKCSJhbGciOiAiYWxnb3JpdGhtIiwKCSJrIjogIm15X2tleSIKICAgIH0sCiAgICAidGVlLWV2aWRlbmNlIjogIm15X2V2aWRlbmNlIgp9Cg==
HTTP/1.1 200 OK
content-length: 0
content-type: application/json
date: Sun, 02 Apr 2023 10:30:07 GMT
```

Do you pull the latest code from main branch?

**Comment by Xynnn007** [comment]
Oh, it seems that you're using a old version of code which has the bug here, and was fixed in https://github.com/confidential-containers/kbs/pull/43/commits/7bdead2fcb63d41116517cd00d2cf9e56268f8ce

**Comment by yaoxin1995** [comment]
> Oh, it seems that you're using a old version of code which has the bug here, and was fixed in [7bdead2](https://github.com/confidential-containers/kbs/commit/7bdead2fcb63d41116517cd00d2cf9e56268f8ce)

Many thanks, I will pull the latest code from main branch and try again:D
